Move thread launching into automotive_demo's main

jwnimmer-tri opened the pull request:
Threads are an application integration concern, not a library concern.  This enables us (in a future PR) to remove thread control virtual methods from the base `DrakeLcmInterface`.

The [full feature branch is here](https://github.com/jwnimmer-tri/drake/tree/lcm-thread-nointerface) for context.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8415)
<!-- Reviewable:end -->

jwnimmer-tri commented:
\CC @caguero @stonier etc. I looked at some downstream code, and I don't _think_ this PR will affect you, but FYI in case you need to adapt any using-code once this commit reaches your builds.

jwnimmer-tri commented:
+@ggould-tri for feature and platform review, please.

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one misgiving below.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[automotive/automotive_demo.cc, line 391 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8415#-L881lFG2LsugGpcyGef:-L881lFG2LsugGpcyGeg:b-ugp6sp) ([raw file](https://github.com/robotlocomotion/drake/blob/e3fa7dd601117335a3e53c8dc79ab312dd596d57/automotive/automotive_demo.cc#L391)):*
> ```cpp
>   const RoadNetworkType road_network_type = DetermineRoadNetworkType();
>   auto simulator = std::make_unique<AutomotiveSimulator<double>>();
>   auto& lcm = dynamic_cast<drake::lcm::DrakeLcm&>(*simulator->get_lcm());
> ```

BTW -- Is there a reason to use a mutable reference here rather than a pointer?  Is it just to get the free RTTI assert from the dynamic cast?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8415#-:-L88381f5FZzLnlBcENx:b-gfhhc3)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[automotive/automotive_demo.cc, line 391 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8415#-L881lFG2LsugGpcyGef:-L884w0d8laPUKhfFHfe:b4n1lb5) ([raw file](https://github.com/robotlocomotion/drake/blob/e3fa7dd601117335a3e53c8dc79ab312dd596d57/automotive/automotive_demo.cc#L391)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW -- Is there a reason to use a mutable reference here rather than a pointer?  Is it just to get the free RTTI assert from the dynamic cast?
</blockquote></details>

Yes, that was why.  Is your worry a nullptr on the way in, or confusing semantics below when `lcm` looks like an object but is really a reference?  I'll tweak this to be clearer / safer depending on what the worry was.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8415)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[automotive/automotive_demo.cc, line 391 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8415#-L881lFG2LsugGpcyGef:-L886zsF977QeyEv9XG2:byhlibx) ([raw file](https://github.com/robotlocomotion/drake/blob/e3fa7dd601117335a3e53c8dc79ab312dd596d57/automotive/automotive_demo.cc#L391)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes, that was why.  Is your worry a nullptr on the way in, or confusing semantics below when `lcm` looks like an object but is really a reference?  I'll tweak this to be clearer / safer depending on what the worry was.
</blockquote></details>

My concern was the object-looking semantics for an unowned mutable; whether the RTTI assert is implicit or explicit is IMO unimportant.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8415)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
CI reminds me that I need to "start the thread" on the DrakeLcmMock before it will publish things for me, so I updated the unit test to obey that for now, with a bunch of one-liner copypasta.  In my feature branch, DrakeLcmMock no longer has a concept of threading (the method will no longer exist), so all of those lines will be deleted.

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[automotive/automotive_demo.cc, line 391 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8415#-L881lFG2LsugGpcyGef:-L88BSw0KBedARBO33ZD:bjw1qmr) ([raw file](https://github.com/robotlocomotion/drake/blob/e3fa7dd601117335a3e53c8dc79ab312dd596d57/automotive/automotive_demo.cc#L391)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

My concern was the object-looking semantics for an unowned mutable; whether the RTTI assert is implicit or explicit is IMO unimportant.
</blockquote></details>

Done.  With my draft edits, cpplint hated the reference cast anyway, so I just changed it to an assertion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8415#-:-L88CaJtzxMoSQkQAaA8:b3qzy8r)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8415)*
<!-- Sent from Reviewable.io -->
